### PR TITLE
FOUR-15854 on task completed - override behavior

### PR DESCRIPTION
**Pull request by devmiguelangel**
## Issue & Reproduction Steps
The user wants to be redirected based on predefined settings, such as the process LaunchPad (of the source process), task list, user homepage dashboard (welcome screen or custom dashboard), a specific dashboard, or an external URL.

## Related Tickets & Packages
[FOUR-15851](https://processmaker.atlassian.net/browse/FOUR-15851)
[FOUR-15854](https://processmaker.atlassian.net/browse/FOUR-15854)
https://github.com/ProcessMaker/screen-builder/pull/1592
https://github.com/ProcessMaker/package-dynamic-ui/pull/141

## Code Review Checklist
- [ ] I have pulled this code locally and tested it on my instance, along with any associated packages.
- [ ] This code adheres to [ProcessMaker Coding Guidelines](https://github.com/ProcessMaker/processmaker/wiki/Coding-Guidelines).
- [ ] This code includes a unit test or an E2E test that tests its functionality, or is covered by an existing test.
- [ ] This solution fixes the bug reported in the original ticket.
- [ ] This solution does not alter the expected output of a component in a way that would break existing Processes.
- [ ] This solution does not implement any breaking changes that would invalidate documentation or cause existing Processes to fail.
- [ ] This solution has been tested with enterprise packages that rely on its functionality and does not introduce bugs in those packages.
- [ ] This code does not duplicate functionality that already exists in the framework or in ProcessMaker.
- [ ] This ticket conforms to the PRD associated with this part of ProcessMaker.

ci:next
ci:modeler:feature/FOUR-15854
ci:screen-builder:feature/FOUR-15854
ci:package-dynamic-u:feature/FOUR-15854

[FOUR-15851]: https://processmaker.atlassian.net/browse/FOUR-15851?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[FOUR-15854]: https://processmaker.atlassian.net/browse/FOUR-15854?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: ProcessMaker/Models/ProcessRequestToken.php [hunk 1]
```diff
@@ -286,29 +288,46 @@ public function assignableUsers()
         return new TokenAssignableUsers($query, $this);
     }
 
-    /**
-     * Get the BPMN definition of the element where the token is.
-     *
-     * @return array|\ProcessMaker\Nayra\Contracts\Bpmn\EntityInterface
-     */
-    public function getDefinition($asObject = false, $par = null)
+    private function getDefinitionFromOwner($asObject)
```
Comment (rodriquelca) [inline ProcessMaker/Models/ProcessRequestToken.php:291]: add the method documentation.

File: ProcessMaker/Models/ProcessRequestToken.php [hunk 2]
```diff
@@ -286,29 +288,46 @@ public function assignableUsers()
...
-        if ($this->getOwner() && $this->getOwnerElement()) {
-            $element = $this->getOwnerElement();
+        $element = $this->getOwnerElement();
 
-            return $asObject ? $element : $element->getProperties();
-        }
+        return $asObject ? $element : $element->getProperties();
+    }
+
+    private function getDefinitionFromRequest($asObject)
```
Comment (rodriquelca) [inline ProcessMaker/Models/ProcessRequestToken.php:298]: add method documentation

File: ProcessMaker/Models/ProcessRequestToken.php [hunk 3]
```diff
@@ -1152,4 +1171,64 @@ public function reassign($toUserId, User $requestingUser)
...
+        try {
+            $elementDestinationProp = json_decode($elementDestinationProp, true);
+            if (is_array($elementDestinationProp) && array_key_exists('type', $elementDestinationProp)) {
+                $elementDestinationType = $elementDestinationProp['type'];
+            }
+        } catch (Exception $e) {
+            return null;
+        }
+
+        switch ($elementDestinationType) {
```
Comment (rodriquelca) [inline ProcessMaker/Models/ProcessRequestToken.php:1196]: Here we can abstract this section into a method example getElementDestinationByType

File: ProcessMaker/Models/ProcessRequestToken.php [hunk 4]
```diff
@@ -1152,4 +1171,64 @@ public function reassign($toUserId, User $requestingUser)
...
+                    } else {
+                        $elementDestination = $elementDestinationProp['value']['url'] ?? null;
+                    }
+                }
+                break;
+            case 'taskList':
+                $elementDestination = route('tasks.index');
+                break;
+            case 'homepageDashboard':
+                if (hasPackage('package-dynamic-ui')) {
```
Comment (rodriquelca) [inline ProcessMaker/Models/ProcessRequestToken.php:1211]: for all element destination route maybe we can create a factory

File: ProcessMaker/Models/ProcessRequestToken.php (hunk 1 above)
Comment (devmiguelangel) [reply ProcessMaker/Models/ProcessRequestToken.php:300]: Doc was added

File: ProcessMaker/Models/ProcessRequestToken.php (hunk 2 above)
Comment (devmiguelangel) [reply ProcessMaker/Models/ProcessRequestToken.php:316]: Doc was added

File: ProcessMaker/Models/ProcessRequestToken.php (hunk 3 above)
Comment (devmiguelangel) [reply ProcessMaker/Models/ProcessRequestToken.php:1197]: The method was improved

File: ProcessMaker/Models/ProcessRequestToken.php (hunk 4 above)
Comment (devmiguelangel) [reply ProcessMaker/Models/ProcessRequestToken.php:1212]: The method was improved

**Review by rodriquelca — APPROVED** [review]

**Pull request merged**